### PR TITLE
Improve on list and add command

### PR DESCRIPTION
**Pull request by limjierui**
Improve on list command to search by dd/mm/yyyy, mm/yyyy, yyyy
Changed date format for add command to dd/mm/yyyy
Add more JunitTests for list command and add command

### CODE REVIEW

File: src/main/java/seedu/mindmymoney/command/ListCommand.java [hunk 1]
```diff
@@ -41,12 +43,12 @@ public boolean isExit() {
     }
 
     /**
-     * Indicates whether the list command is to list expenditure(s) by looking for the /expenses flag.
+     * Indicates whether the list command is to list expenditure(s) by looking for the /pm flag.
```
Comment (SeanHoWB) [inline src/main/java/seedu/mindmymoney/command/ListCommand.java:46]: shld be /e

File: src/main/java/seedu/mindmymoney/command/ListCommand.java [hunk 2]
```diff
@@ -67,25 +69,71 @@ private boolean hasIncomeListFlag() {
...
      * Gets all expenditures and formats them into a String to be printed.
      *
      * @return String of expenditures.
+     * @throws MindMyMoneyException Throws an exception when the date is not in the correct format
      */
-    public String expenditureListToString() {
+    public String expenditureListToString() throws MindMyMoneyException {
         int indexOfList = 1;
         String listInString = "";
+        if (listInput.equals(FLAG_OF_EXPENSES)) {
```
Comment (SeanHoWB) [inline src/main/java/seedu/mindmymoney/command/ListCommand.java:81]: should avoid arrowhead code

File: src/main/java/seedu/mindmymoney/helper/Calculations.java [hunk 3]
```diff
@@ -45,10 +47,26 @@ public static void calculateExpenditurePerMonth(String input, ExpenditureList ex
...
         displayExpenditureBreakdown(foundItems, sumOfExpenditure);
     }
 
+    /**
+     * Checks if date input format is valid.
+     *
+     * @param input The string of the date input.
+     * @return true if format is valid, false otherwise.
+     */
+    public static boolean isValidInput(String input) {
```
Comment (SeanHoWB) [inline src/main/java/seedu/mindmymoney/helper/Calculations.java:60]: can consider putting in general functions/ time functiuons

File: src/test/java/seedu/mindmymoney/command/ListCommandTest.java [hunk 4]
```diff
@@ -12,26 +12,99 @@
...
         User user = new User(expenditureTestList, creditCardTestList, incomeList);
-
-        String inputString = "/pm cash /c Personal /d Nike Shoes /a 300 /t 2022-03";
+        String inputString = "/pm cash /c Personal /d Nike Shoes /a 300 /t 30/03/2022";
         new AddCommand(inputString, user).executeCommand();
         String listInString = new ListCommand("/e", user).expenditureListToString();
-        assertEquals("1. $300.0 on Nike Shoes. Paid using Cash [Personal] [Mar 2022]\n", listInString);
+        assertEquals("1. $300.0 was spent on Nike Shoes(Personal) using Cash [30/03/2022]\n", listInString);
+
+        String inputString2 = "/pm cash /c Food /d Cream Pie /a 69 /t 30/03/2022";
```
Comment (SeanHoWB) [inline src/test/java/seedu/mindmymoney/command/ListCommandTest.java:29]: better naming of the inputstring?

Comment (SeanHoWB) [reply src/test/java/seedu/mindmymoney/command/ListCommandTest.java:29]: secondInputString perhaps, or specify what its for

**Review by SeanHoWB — COMMENTED** [review]
not bad! tidy up abit and itll be good for merging!
image

File: src/main/java/seedu/mindmymoney/command/ListCommand.java [hunk 5]
```diff
@@ -67,25 +69,71 @@ private boolean hasIncomeListFlag() {
...
+        assert listInString.length() != 0 : "Return string should be non-empty";
+        return listInString;
+    }
 
+    /**
+     * Formats the output of expenses in list.
+     *
+     * @param index To obtain the numbering when listing the expenses.
+     * @param listInString String where the content of output is appended to.
+     * @return
```
Comment (khseah) [inline src/main/java/seedu/mindmymoney/command/ListCommand.java:110]: return missing

File: src/main/java/seedu/mindmymoney/command/ListCommand.java [hunk 6]
```diff
@@ -67,25 +69,71 @@ private boolean hasIncomeListFlag() {
...
+                        listInString += indexOfList + ". $" + i.getAmount() + " was spent on " + i.getDescription()
+                                + "(" + i.getCategory() + ") " + "using " + i.getPaymentMethod()
+                                + " [" + i.getTime() + "]" + "\n";
+                        indexOfList++;
+                    }
+                }
+            } else {
+                listInString = printListString(indexOfList, listInString);
+            }
+        }
```
Comment (khseah) [inline src/main/java/seedu/mindmymoney/command/ListCommand.java:100]: Consider extracting out code to improve SLAP and avoid "arrowhead" style

File: src/main/java/seedu/mindmymoney/command/ListCommand.java [hunk 7]
```diff
@@ -67,25 +69,71 @@ private boolean hasIncomeListFlag() {
...
+     * @return true if format is valid, false otherwise.
+     */
+    public static boolean isValidInput(String input) {
+        if (input.matches(ValidationRegexTypes.VALIDATION_REGEX_D)
+                || input.matches(ValidationRegexTypes.VALIDATION_REGEX_M)
+                || input.matches(ValidationRegexTypes.VALIDATION_REGEX_Y)) {
+            return true;
+        } else {
+            return false;
+        }
```
Comment (khseah) [inline src/main/java/seedu/mindmymoney/command/ListCommand.java:134]: Minor thing, but can just return false
```suggestion
        return false;
```

**Review by khseah — COMMENTED** [review]
Nicely done

File: src/main/java/seedu/mindmymoney/command/AddCommand.java [hunk 8]
```diff
@@ -116,6 +117,20 @@ public void addExpenditure() throws MindMyMoneyException {
         System.out.print(System.lineSeparator());
     }
 
+    /**
+     * Checks if date input format is valid.
+     *
+     * @param input The string of the date input.
+     * @return true if format is valid, false otherwise.
+     */
+    public static boolean isValidInput(String input) {
```
[the remark this answers is not in the document]
Comment (khseah) [reply src/main/java/seedu/mindmymoney/command/AddCommand.java:126]: HAHAHAHA the picture made my day

[the remark this answers is not in the document]
Comment (limjierui) [reply src/main/java/seedu/mindmymoney/command/AddCommand.java:126]: resolved. Thanks for the picture hahaha

File: src/main/java/seedu/mindmymoney/command/ListCommand.java (hunk 1 above)
Comment (limjierui) [reply src/main/java/seedu/mindmymoney/command/ListCommand.java:46]: resolved

File: src/main/java/seedu/mindmymoney/command/ListCommand.java (hunk 2 above)
Comment (limjierui) [reply src/main/java/seedu/mindmymoney/command/ListCommand.java:81]: resolved

File: src/main/java/seedu/mindmymoney/command/ListCommand.java (hunk 5 above)
Comment (limjierui) [reply src/main/java/seedu/mindmymoney/command/ListCommand.java:110]: resolved

File: src/main/java/seedu/mindmymoney/command/ListCommand.java (hunk 7 above)
Comment (limjierui) [reply src/main/java/seedu/mindmymoney/command/ListCommand.java:134]: resolved

File: src/main/java/seedu/mindmymoney/helper/Calculations.java [hunk 9]
```diff
@@ -45,10 +47,26 @@ public static void calculateExpenditurePerMonth(String input, ExpenditureList ex
...
+        if (input.matches(ValidationRegexTypes.VALIDATION_REGEX_D)
+                || input.matches(ValidationRegexTypes.VALIDATION_REGEX_M)
+                || input.matches(ValidationRegexTypes.VALIDATION_REGEX_Y)) {
+            return true;
+        } else {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (limjierui) [reply src/main/java/seedu/mindmymoney/helper/Calculations.java:67]: resolved

File: src/test/java/seedu/mindmymoney/command/ListCommandTest.java [hunk 10]
```diff
@@ -12,26 +12,99 @@
...
+        new AddCommand(inputString, user).executeCommand();
+        String listInString = new ListCommand("/e 2022", user).expenditureListToString();
+        assertEquals("1. $300.0 was spent on Nike Shoes(Personal) using Cash [30/03/2022]\n", listInString);
+        String inputString2 = "/pm cash /c Food /d Cream Pie /a 69 /t 30/03/2022";
+        new AddCommand(inputString2, user).executeCommand();
+        String inputString3 = "/pm cash /c Food /d Cream Pie /a 69 /t 30/04/2021";
+        new AddCommand(inputString3, user).executeCommand();
+        listInString = new ListCommand("/e 2022", user).expenditureListToString();
+        assertEquals("1. $300.0 was spent on Nike Shoes(Personal) using Cash [30/03/2022]\n"
+                + "2. $69.0 was spent on Cream Pie(Food) using Cash [30/03/2022]\n", listInString);
```
[the remark this answers is not in the document]
Comment (limjierui) [reply src/test/java/seedu/mindmymoney/command/ListCommandTest.java:107]: resolved

**Pull request merged**